pkg/didcomm/common/service/service.go

```diff
@@ -20,16 +27,55 @@ type DIDComm interface {
 	Event
 }
 
+// Header helper structure which keeps reusable fields
+type Header struct {
+	ID     string           `json:"@id"`
+	Thread decorator.Thread `json:"~thread"`
+	Type   string           `json:"@type,omitempty"`
+}
+
 // DIDCommMsg did comm msg
 type DIDCommMsg struct {
 	// Outbound indicates the direction of this DIDComm message:
 	//   - outgoing (to another agent)
 	//   - incoming (from another agent)
 	Outbound bool
-	Type     string
 	Payload  []byte
 	// TODO : might need refactor as per the issue-226
 	OutboundDestination *Destination
+	Header              *Header
+}
+
+// NewDIDCommMsg returns DIDCommMsg with Header
+func NewDIDCommMsg(payload []byte) (*DIDCommMsg, error) {
+	msg := &DIDCommMsg{Payload: payload}
+	if err := json.Unmarshal(msg.Payload, &msg.Header); err != nil {
+		return nil, fmt.Errorf("invalid payload data format: %w", err)
+	}
+	return msg, nil
+}
+
+// ThreadID returns msg ~thread.thid if there is no ~thread.thid returns msg @id
+// message is invalid if ~thread.thid exist and @id is absent
+func (m *DIDCommMsg) ThreadID() (string, error) {
+	if m.Header == nil {
```

Comment:
When would this happen?

Reply:
when someone will create `DIDCommMsg{}` instead of using `NewDIDCommMsg(src)`

Reply:
May be disclaimer in the GoDoc would be helpful imo